File: go/classes/chart_maker.go
// Chart Making — a visualization pipeline as composed functions.
package classes

import (
	"fmt"
	"math"
	"strings"
)

// DataPoint is one observation with a label and a numeric value.
type DataPoint struct {
	Label  string
	Value  float64
	Series string
}

// LinearScale maps a numeric domain to a numeric range.
type LinearScale struct {
	DomainMin, DomainMax float64
	RangeMin, RangeMax   float64
}

// At evaluates the scale at x.
func (s LinearScale) At(x float64) float64 {
	if s.DomainMax == s.DomainMin {
		return (s.RangeMin + s.RangeMax) / 2.0
	}
	t := (x - s.DomainMin) / (s.DomainMax - s.DomainMin)
	return s.RangeMin + t*(s.RangeMax-s.RangeMin)
}

// FitLinearScale builds a scale that comfortably spans the provided values.
func FitLinearScale(values []float64, rangeMin, rangeMax float64, includeZero bool) LinearScale {
	if len(values) == 0 {
		return LinearScale{DomainMin: 0, DomainMax: 1, RangeMin: rangeMin, RangeMax: rangeMax}
	}
	lo := math.Inf(1)
	hi := math.Inf(-1)
	for _, v := range values {
		if v < lo {
			lo = v
		}
		if v > hi {
			hi = v
		}
	}
	if includeZero {
		lo = math.Min(lo, 0)
		hi = math.Max(hi, 0)
	}
	if lo == hi {
		lo -= 1
		hi += 1
	}
	return LinearScale{DomainMin: lo, DomainMax: hi, RangeMin: rangeMin, RangeMax: rangeMax}
}

// OrdinalScale maps discrete categories to consecutive positions.
type OrdinalScale struct {
	Categories         []string
	RangeMin, RangeMax int
}

// At returns the position for a category, or -1 if unknown.
func (s OrdinalScale) At(category string) int {
	idx := -1
	for i, c := range s.Categories {
		if c == category {
			idx = i
			break
		}
	}
	if idx < 0 {
		return -1
	}
	n := len(s.Categories)
	if n <= 1 {
		return (s.RangeMin + s.RangeMax) / 2
	}
	step := float64(s.RangeMax-s.RangeMin) / float64(n-1)
	return int(math.Round(float64(s.RangeMin) + float64(idx)*step))
}

// BarChart is a horizontal ASCII bar chart spec.
type BarChart struct {
	Title  string
	Points []DataPoint
	Width  int
}

// Render produces the chart as a string.
func (c BarChart) Render() string {
	if len(c.Points) == 0 {
		return c.Title + "\n(no data)"
	}
	values := make([]float64, len(c.Points))
	for i, p := range c.Points {
		values[i] = p.Value
	}
	scale := FitLinearScale(values, 0, float64(c.Width), true)
	labelW := 0
	for _, p := range c.Points {
		if len(p.Label) > labelW {
			labelW = len(p.Label)
		}
	}
	var b strings.Builder
	b.WriteString(c.Title)
	b.WriteByte('\n')
	b.WriteString(strings.Repeat("=", labelW+4+c.Width+10))
	b.WriteByte('\n')
	for _, p := range c.Points {
		barLen := int(math.Round(scale.At(p.Value)))
		if barLen < 0 {
			barLen = 0
		}
		bar := strings.Repeat("█", barLen)
		fmt.Fprintf(&b, "  %-*s  %-*s  %.2f\n", labelW, p.Label, c.Width, bar, p.Value)
	}
	return strings.TrimRight(b.String(), "\n")
}

// LineChart is a vertical ASCII line-chart spec.
type LineChart struct {
	Title          string
	Points         []DataPoint
	Width, Height  int
}

// Render produces the chart as a string.
func (c LineChart) Render() string {
	if len(c.Points) == 0 {
		return c.Title + "\n(no data)"
	}
	n := len(c.Points)
	values := make([]float64, n)
	for i, p := range c.Points {
		values[i] = p.Value
	}
	xDomainMax := float64(n - 1)
	if xDomainMax < 1 {
		xDomainMax = 1
	}
	xScale := LinearScale{DomainMin: 0, DomainMax: xDomainMax,
		RangeMin: 0, RangeMax: float64(c.Width - 1)}
	yScale := FitLinearScale(values, 0, float64(c.Height-1), false)

	grid := make([][]rune, c.Height)
	for i := range grid {
		grid[i] = make([]rune, c.Width)
		for j := range grid[i] {
			grid[i][j] = ' '
		}
	}
	plot := func(col, row int, r rune) {
		if col >= 0 && col < c.Width && row >= 0 && row < c.Height {
			grid[row][col] = r
		}
	}

	var pc, pr int
	var havePrev bool
	for i, p := range c.Points {
		col := int(math.Round(xScale.At(float64(i))))
		row := c.Height - 1 - int(math.Round(yScale.At(p.Value)))
		plot(col, row, '●')
		if havePrev {
			dx := col - pc
			dy := row - pr
			steps := int(math.Max(math.Abs(float64(dx)), math.Abs(float64(dy))))
			if steps > 1 {
				for s := 1; s < steps; s++ {
					t := float64(s) / float64(steps)
					ic := int(math.Round(float64(pc) + float64(dx)*t))
					ir := int(math.Round(float64(pr) + float64(dy)*t))
					if ir >= 0 && ir < c.Height && ic >= 0 && ic < c.Width && grid[ir][ic] == ' ' {
						plot(ic, ir, '·')
					}
				}
			}
		}
		pc, pr = col, row
		havePrev = true
	}

	var b strings.Builder
	b.WriteString(c.Title)
	b.WriteByte('\n')
	hi := yScale.DomainMax
	lo := yScale.DomainMin
	for r := 0; r < c.Height; r++ {
		var left string
		switch r {
		case 0:
			left = fmt.Sprintf("%7.2f ┤", hi)
		case c.Height - 1:
			left = fmt.Sprintf("%7.2f ┤", lo)
		default:
			left = "        │"
		}
		b.WriteString(left)
		b.WriteString(string(grid[r]))
		b.WriteByte('\n')
	}
	b.WriteString("        └")
	b.WriteString(strings.Repeat("─", c.Width))
	return b.String()
}

File: python/classes/chart_maker.py
"""Chart Making — a visualization pipeline as composed functions.

The chart is not a container of drawing commands. It is a small pipeline:

  data → scale → encode → layout → render

Each stage is a pure function. `LinearScale(d_min, d_max, r_min, r_max)` is
literally a function from a data domain to a visual range. `Encoding` is the
binding between data fields and visual channels (position, size, label).
`render` is the pluggable backend — here ASCII, but the same spec could emit
SVG or canvas with a different renderer.

This is the grammar-of-graphics shape: declarative spec + rendering backend,
separating WHAT TO SHOW from HOW TO DRAW IT.
"""

from __future__ import annotations
from dataclasses import dataclass, field
from typing import Callable


@dataclass
class DataPoint:
    label: str
    value: float
    series: str = ""


@dataclass
class Series:
    name: str
    points: list[DataPoint]


@dataclass
class LinearScale:
    """A function from a numeric domain to a numeric range. Callable."""
    domain_min: float
    domain_max: float
    range_min: float
    range_max: float

    def __call__(self, x: float) -> float:
        if self.domain_max == self.domain_min:
            return (self.range_min + self.range_max) / 2
        t = (x - self.domain_min) / (self.domain_max - self.domain_min)
        return self.range_min + t * (self.range_max - self.range_min)

    @staticmethod
    def fit(values: list[float], range_min: float, range_max: float,
            include_zero: bool = True) -> LinearScale:
        """Build a scale that comfortably spans the data."""
        if not values:
            return LinearScale(0, 1, range_min, range_max)
        lo = min(values)
        hi = max(values)
        if include_zero:
            lo = min(lo, 0)
            hi = max(hi, 0)
        if lo == hi:
            lo -= 1
            hi += 1
        return LinearScale(lo, hi, range_min, range_max)


@dataclass
class OrdinalScale:
    """Maps a discrete set of categories to consecutive positions."""
    categories: list[str]
    range_min: int
    range_max: int

    def __call__(self, category: str) -> int:
        if category not in self.categories:
            raise ValueError(f"unknown category: {category}")
        n = len(self.categories)
        idx = self.categories.index(category)
        if n <= 1:
            return (self.range_min + self.range_max) // 2
        step = (self.range_max - self.range_min) / (n - 1)
        return int(self.range_min + idx * step)


# --- chart specifications ---

@dataclass
class BarChart:
    """Horizontal ASCII bar chart."""
    title: str
    points: list[DataPoint]
    width: int = 40

    def render(self) -> str:
        if not self.points:
            return f"{self.title}\n(no data)"
        values = [p.value for p in self.points]
        # Bars grow from 0 to max; negatives aren't modeled here.
        scale = LinearScale.fit(values, 0, self.width, include_zero=True)
        label_w = max(len(p.label) for p in self.points)
        lines = [self.title, "=" * (label_w + 4 + self.width + 10)]
        for p in self.points:
            bar_len = max(0, int(round(scale(p.value))))
            bar = "█" * bar_len
            lines.append(f"  {p.label.ljust(label_w)}  {bar.ljust(self.width)}  {p.value:.2f}")
        return "\n".join(lines)


@dataclass
class LineChart:
    """Vertical ASCII line chart. Points plotted on a 2D grid of characters."""
    title: str
    points: list[DataPoint]
    width: int = 40
    height: int = 12

    def render(self) -> str:
        if not self.points:
            return f"{self.title}\n(no data)"
        n = len(self.points)
        values = [p.value for p in self.points]
        x_scale = LinearScale(0, max(1, n - 1), 0, self.width - 1)
        y_scale = LinearScale.fit(values, 0, self.height - 1, include_zero=False)
        grid = [[" "] * self.width for _ in range(self.height)]

        def plot(col: int, row: int, ch: str) -> None:
            if 0 <= col < self.width and 0 <= row < self.height:
                grid[row][col] = ch

        # Plot each point and a crude line segment to the previous one.
        prev: tuple[int, int] | None = None
        for i, p in enumerate(self.points):
            col = int(round(x_scale(i)))
            # y axis grows down; invert row index.
            row = self.height - 1 - int(round(y_scale(p.value)))
            plot(col, row, "●")
            if prev is not None:
                pc, pr = prev
                # Simple line: step along the longer axis.
                dx = col - pc
                dy = row - pr
                steps = max(abs(dx), abs(dy))
                if steps > 1:
                    for s in range(1, steps):
                        t = s / steps
                        ic = int(round(pc + dx * t))
                        ir = int(round(pr + dy * t))
                        if grid[ir][ic] == " ":
                            plot(ic, ir, "·")
            prev = (col, row)

        # Left axis with max/min labels.
        hi = y_scale.domain_max
        lo = y_scale.domain_min
        out = [self.title]
        for r in range(self.height):
            left = ""
            if r == 0:
                left = f"{hi:>7.2f} ┤"
            elif r == self.height - 1:
                left = f"{lo:>7.2f} ┤"
            else:
                left = "        │"
            out.append(left + "".join(grid[r]))
        out.append("        └" + "─" * self.width)
        xticks_line = ["        "]
        # Show first and last labels (others would crowd).
        first_label = self.points[0].label
        last_label = self.points[-1].label
        padding = " " * max(0, self.width - len(first_label) - len(last_label) - 1)
        out.append("         " + first_label + padding + last_label)
        return "\n".join(out)


# --- demo ---
if __name__ == "__main__":
    # Monthly revenue → bar chart
    monthly = [
        DataPoint("Jan", 12000), DataPoint("Feb", 15500),
        DataPoint("Mar", 18700), DataPoint("Apr", 14200),
        DataPoint("May", 21300), DataPoint("Jun", 19800),
    ]
    print(BarChart(title="Monthly Revenue ($)", points=monthly, width=30).render())
    print()

    # Daily temperatures → line chart
    temps = [
        DataPoint(f"D{i}", v) for i, v in
        enumerate([15.2, 16.1, 18.5, 22.0, 24.3, 23.7, 20.1,
                   17.8, 19.4, 21.9, 23.2, 25.1, 22.6, 18.3])
    ]
    print(LineChart(title="Two-Week Temperature (°C)", points=temps,
                    width=40, height=10).render())

    # Compose: the same data can feed either chart type.
    # The spec (which chart) is declarative; the renderer is pluggable.
    print()
    print("(same 14-day data as bars, truncated to first 7):")
    print(BarChart(title="First Week Temp (°C)", points=temps[:7], width=25).render())
